src/optimization/steepest_descent_method.cpp

```diff
@@ -0,0 +1,302 @@
+ {
+	communicator->send(target, k52::parallel::mpi::constants::kCommonTag, increment_of_the_argument_);
+	communicator->send(target, k52::parallel::mpi::constants::kCommonTag, max_iteration_number_);
+	communicator->send(target, k52::parallel::mpi::constants::kCommonTag, precision_);
+}
+
+void SteepestDescentMethod::Receive(boost::mpi::communicator* communicator, int source)
+{
+	communicator->recv(source, k52::parallel::mpi::constants::kCommonTag, increment_of_the_argument_);
+	communicator->recv(source, k52::parallel::mpi::constants::kCommonTag, max_iteration_number_);
+	communicator->recv(source, k52::parallel::mpi::constants::kCommonTag, precision_);
+}
+#endif
+
+
+vector<double> SteepestDescentMethod::FindOptimalParameters(const vector<double>& initial_parameters)
+{
+	vector<double> parameters(initial_parameters);
+	size_t counter = 0;
+	double lambda = 1; 
+
+	do
+	{
+		counter++;
+		vector<double> gradient = CalculateGradient(parameters);
+		vector<double> invert_gradiend = InvertVector(gradient);
+
+		lambda = GoldenSectionSearch(parameters, lambda, invert_gradiend);
+
+		vector<double> next_step_parameters(parameters.size());
+		for (size_t i = 0; i < next_step_parameters.size(); ++i)
+		{
+			next_step_parameters[i] = parameters[i] + lambda * invert_gradiend[i];
+		}
+
+		if (IsExitCriteriaFulfilled(next_step_parameters))
+		{
+			break;
+		}
+		
+		parameters = next_step_parameters;
+
+		
+		if (counter == max_iteration_number_)
+		{
+			std::cout << " Minimun has not found"<<std::endl;
+			break;
+		}
+	} while (true);
+
+	return parameters;
+}
+
+bool SteepestDescentMethod::IsExitCriteriaFulfilled(
+	vector<double>& steps_array) const
+{
+	double temp = 0;
+	vector<double> gradient = CalculateGradient(steps_array);
+	for (size_t i = 0; i < steps_array.size(); ++i)
+	{
+		temp += gradient[i] * gradient[i];
+	}
+	if (sqrt(temp) > precision_)
+	{
+		return false;
+	}
+	return true;
+}
+
+vector<double> SteepestDescentMethod::CalculateGradient(
+	const vector<double>& parameters) const
+{
+	vector<double> gradient(parameters.size());
+
+	for (size_t i = 0; i<parameters.size(); ++i)
+	{
+		gradient[i] = CalculateDerivative(parameters, i);
+	}
+
+	return gradient;
+}
+
+double SteepestDescentMethod::CalculateDerivative(
+	const vector<double>& parameters,
+	size_t& index) const
+{
+	vector<double> decrement_function(parameters);
+	vector<double> increment_function(parameters);
+	decrement_function[index] = parameters[index] - increment_of_the_argument_ / 2;
+	increment_function[index] = parameters[index] + increment_of_the_argument_ / 2;
+	double increment_function_value = CountObjectiveFunctionValueToMinimize(increment_function);
+	double decrement_function_value = CountObjectiveFunctionValueToMinimize(decrement_function);
+	return (increment_function_value - decrement_function_value) / increment_of_the_argument_;
+}
+
+vector<double> SteepestDescentMethod::InvertVector(
+	vector<double>& parameters) const
+{
```

Comment:
Move to k52::common::Math as static method (create this class, see FloatingPoint as example) and implement unit tests for this metod.